src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -347,6 +418,14 @@ async def coursier_resolve_lockfile(
     if len(artifact_requirements) == 0:
         return CoursierResolvedLockfile(entries=())
 
+    # Transform requirements with local JAR files into coordinates, and yoink the relevant
+    # files into the coursier sandbox
+    better_artifacts = await Get(
+        ArtifactRequirementsWithLocalFiles, ArtifactRequirements, artifact_requirements
+    )
```

Comment:
I think that scanning to find all `jvm_artifact` URLs, and then matching them with coordinates like this isn't ideal.

Instead, what about 1) adjusting `ArtifactRequirements` to be a collection of a new class `CoordinateInput` (or `CoordinateWithURL` etc), or 2) adding a `url` to `Coordinate`? I think that I would lean slightly toward approach 1, because `Coordinate` is frequently used as the _output_ of a resolve, and in that case the URL would never be present.

Reply:
I'm confused. `Coordinate` has URLs (per my previous PR), and this wasn't caught there. Is this a change you want more generally?

Reply:
Ah, true: in that case, adding a `jar: Digest` field to `Coordinate` could be ok. The general thought on inputs vs outputs might still apply, it just didn't occur to me until thinking about an alternative to the replacement that you're doing here.

Reply:
@stuhood In which case, would you object to landing the PR as it currently stands, and I'll do the input/output split as a fast-follow?

Reply:
That might make sense if this were already green, but since it isn't, and there is more pending feedback, it seems worthwhile to do it here...? Or alternatively, as prework?

Reply:
@stuhood The CI errors aren't related to my code, they're related to https://pantsbuild.slack.com/archives/C0D7TNJHL/p1638933581205900 and all remaining feedback is resolved